### PR TITLE
Support zero dimensions in softmax decompositions

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #91322

The eager implementation of softmax supports computation along zero dimensions, but many of the other implementations did not, including:
* decompositions & refs (this was causing dynamo failures)
* forward AD for logsumexp
* MPS log_softmax_backward

This PR handles the `input.numel() == 0` cases separately to avoid running `amax()`, which fails for zero dimensions, and updates opinfos.

example of "computation along zero dimensions":

```python
# example of where 
import torch

t = torch.rand((4, 0, 0))
print("~")
print(torch.nn.functional.softmax(t, dim=-1))  # this passes
print("~")
torch._refs.softmax(t, dim=-1)  # this fails
print("~")
```

### CODE REVIEW

**Comment by davidberard98** [comment]
cc @lezcano who I think has been doing related work recently?

**Comment by davidberard98** [comment]
nvm, looks like the mps failure is legitimate; going to take a look at that first

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -987,6 +987,8 @@ def _softmax(x: Tensor, dim: int, half_to_float: bool):
     computation_dtype, result_dtype = utils.elementwise_dtypes(
         x, type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT
     )
+    if x.numel() == 0:
+        return torch.zeros_like(x, dtype=result_dtype)
```
Comment (lezcano) [inline torch/_decomp/decompositions.py:991]: See that the cast to `result_type` just happens when `not half_to_float`.

To avoid these issues, I would suggest to simply set `x_max = zeros_like(x)` in this case, rather than trying to completely short cut the computation in this case.

**Review by lezcano — COMMENTED** [review]
A small point.

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (davidberard98) [reply torch/_decomp/decompositions.py:991]: thanks! let me update that.

**Review by lezcano — APPROVED** [review]
LGTM!
